Pretty-print parenthesis around binary in postfix match

scrabsha opened the pull request:
Fixes #124206.

scrabsha committed:
Pretty-print parenthesis around binary in postfix match

Signed-off-by: Sasha Pourcelot <sasha.pourcelot@protonmail.com>

scrabsha commented on compiler/rustc_ast_pretty/src/pprust/state/expr.rs:
```diff
@@ -488,7 +488,18 @@ impl<'a> State<'a> {
                         self.space();
                     }
                     MatchKind::Postfix => {
+                        let need_par = matches!(expr.kind, ast::ExprKind::Binary(_, _, _));
```
I looked at the [Expression precedence][1] table of the Reference and found nothing else to wrap in parentheses.

Please comment if I missed something!

[1]: https://doc.rust-lang.org/reference/expressions.html#expression-precedence

compiler-errors replied:
It's probably best to use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/struct.Expr.html#method.precedence instead of matching exhaustively on `ExprKind`

like `.precedence() <= PREC_SOMETHING` or `<` or something (idk which predecence to choose)

scrabsha replied:
Thanks for your comment. This allowed me to fix a few other instances of the bug reported by #124206!

RossSmyth commented:
Could you please either give the test name something descriptive about what the test tests (idk pfix-match-paren or something) or make a new directory for postfix match within the pretty directory. Adding tests with issue numbers, while common, makes maintenance frustrating.

scrabsha commented:
> Could you please either give the test name something descriptive about what the test tests (idk pfix-match-paren or something) or make a new directory for postfix match within the pretty directory. Adding tests with issue numbers, while common, makes maintenance frustrating.

I `mkdir`ed `tests/pretty/postfix-match` and moved every postfix-match-related test in it. Thanks for the suggestion!

scrabsha commented on tests/pretty/postfix-match/precedence.pp:
```diff
@@ -0,0 +1,25 @@
+
+pub fn main() {
+    let a;
+    (
+        { 1 } + 1).match {
+        _ => {}
+    };
```
This looks incorrect. Is it an issue in my code or in the pretty printer?

dtolnay reviewed (changes requested):
This is going to cause double parens to be put around cases like `repro!(1 + Struct {})`. `((1 + Struct {})).match { _ => {} }`

In general `print_expr_as_cond` is not a correct thing to be using here. That is for syntactic positions where a `{` is ambiguous between being a part of the matched expression / if condition / while condition, vs being the body for the match/if/while. It isn't applicable to postfix match.

Postfix match needs to be printed like very other expression kind containing `.`, such as field accesses and method call expressions and postfix await.

scrabsha commented:
@dtolnay I modified my changes according to your comment and added your example in the test suite. Thanks!

@rustbot label -S-waiting-on-author +S-waiting-on-review

dtolnay reviewed (approved):
Thanks!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit c8ff8a4dc7c9867c54df8f6d31120f4cd3923508 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay c8ff8a4dc7c9867c54df8f6d31120f4cd3923508 -->
<!-- homu: {"type":"Approved","sha":"c8ff8a4dc7c9867c54df8f6d31120f4cd3923508","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
